### PR TITLE
Register an Olympus application for push notification #613

### PR DESCRIPTION
**Pull request by ognjenns**

### CODE REVIEW

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.m]: Please check typo: getCredenitalsSid

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (atsakiridis) [inline Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm]: My understanding is that this ends up doing asynchronous operations (i.e. HTTP provisioning requests for the various push related entities), shouldn't we pass a delegate and be able to get a callback when the provisioning requests are finished properly, or an error if something went wrong?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.m]: Hmm, I'd prefer a different name. PushHandler doesn't seem to accurately describe what this class does, which is provisioning the needed entities for push in the server side, right? Maybe something like PushProvisioningHandler?

File: RestCommClient/Classes/Push/PushHandler.h
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.h]: Let's make it a habit to add some rough doc for new entities even if they aren't directly user facing API. Just couple of lines of what it does. At some point later we want to be able to generate Appledoc for any of our modules to assist community and newcomers in the team, as well as reviewers ;)

File: RestCommClient/Classes/Push/PushCommunicationManager.h
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushCommunicationManager.h]: Same as PushHandler, let's add some doc here

File: RestCommClient/Classes/Push/Binding.h
Comment (atsakiridis) [inline RestCommClient/Classes/Push/Binding.h:23]: Same as PushHandler, let's add some doc here

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.m:160]: Is there a chance that the accountSid exists in the 'cache' but is not valid and we need to handle that case?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.m:87]: I'm worried that all the complexity introduced by all those block + completion handlers will come back to bite us in the future in terms of bugs and difficulty to troubleshoot issues. Could we maybe use synchronous http facilities and run on after the other instead, while all in a background thread of course so that we don't mess with UI thread? That could potentially make error reporting easier as well. Wdyt?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushHandler.m:1]: Have you considered introducing testing facilities for the PushHandler (provisioning) functionality? Seems like a good candidate

**Review by atsakiridis — CHANGES_REQUESTED** [review]

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m:1]: Nope. If we need to introduce testing, I will need to organize the code differently. I am for it if we have time.

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m]: The idea is to use it for all the push stuff we have in the future, so thats why I want it only to be "Handler"; We can have it different provisioning in IOS context, for me, have different meaning. Maybe we can leave it as it is, or something different than provisioning? any suggestions?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m:87]: I was start thinking in that way, but the only thing I have found, using the NSURLSession, is to use semaphore example: https://stackoverflow.com/questions/26784315/can-i-somehow-do-a-synchronous-http-request-via-nsurlsession-in-swift
And I saw "wait forever"

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m]: Thanks

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m:160]: Good point. Solution will be to have combination saved:
accountSid and rescommAccountEmail; in this case, we will be sure that we have saved the account sid for email. Also, we must ask server side is the account sid persistant. wdyt?

File: RestCommClient/Classes/Push/PushHandler.h
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.h]: Ok, I agree. Will add it.

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (ognjenns) [reply Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm]: I thought to change the SDK as less as possible. But, at the end, you are right. As A developer I would like to know what happened :)

Comment (atsakiridis) [reply Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm]: We want to change as little as possible in terms of existing API being backward compatible. This is a new API, so let's make sure its as usable as possible

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (atsakiridis) [inline Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm:223]: I think we need to start getting into the habit of adding proper constants for key names for API related dictionaries. Can you provide such things for keys like 'friendly-name', 'username', etc?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushHandler.m]: Ok if you plan to add more push related functionality, let's wait until this functionality lands and we can re-review then

File: RestCommClient/Classes/Push/PushCommunicationManager.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushCommunicationManager.m]: Name PushCommuniationManager seems weird here, like it's about actual push communication messages management, while we are talking about configuration/provisioning, right? What about something like PushConfigurationClient? Maybe also move caching functionality from PushHandler here, so that in the future we can integration test the whole client functionality in isolation. Wdyt?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushHandler.m:1]: Let's do it as discussed in slack

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushHandler.m:87]: You don't have to wait forever, you can use a timeout instead. The idea here is that we use tools that are fit for the job. Async is great in most cases, but right now I'm getting the impression that we are doing serial processing and with so many asynchronous operations + blocks we might be asking for trouble. In any case I don't feel that strongly about it; just want to make sure we know where we 're getting, so I'll let you decide

File: RestCommClient/Classes/Push/Binding.h
Comment (atsakiridis) [inline RestCommClient/Classes/Push/Binding.h:26]: How come we have Binding.h/m but no similar modules for account, credentials, etc?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushHandler.m:160]: I don't have full picture on this implementation-wise, so use your best judgement. Just wanted to point out to make sure that cache stays up to date during configuration changes

**Review by atsakiridis — CHANGES_REQUESTED** [review]

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m]: ok

File: RestCommClient/Classes/Push/Binding.h
Comment (ognjenns) [reply RestCommClient/Classes/Push/Binding.h:26]: we don't need to have all the entities saved on SDK side.
All we need are Sid's. 
Binding object is something we create/update and we need all data just for it.

File: RestCommClient/Classes/Push/PushHandler.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushHandler.m:87]: I think we are good. I would like to leave it as it is.

File: RestCommClient/Classes/Push/PushCommunicationManager.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushCommunicationManager.m]: Maybe we can call PushApiManager? We actually calling API (server) to obtain some stuff.
Therefore, I would like to not cache anything in it. To have only api calls. And then, when we introduce testing, we can easily mock it? Wdyt?

File: RestCommClient/Classes/Push/PushHandler.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushHandler.m:87]: Ok, let's hope it doesn't come back to bite us

File: RestCommClient/Classes/Push/PushCommunicationManager.m
Comment (atsakiridis) [reply RestCommClient/Classes/Push/PushCommunicationManager.m]: Heheh, ok let's do it like that. We can re-discuss if needed when the whole thing is ready and we can get better sense of the whole thing

File: RestCommClient/Classes/Push/Binding.h
Comment (atsakiridis) [reply RestCommClient/Classes/Push/Binding.h:26]: Don't you also create account and credential entities?

Comment (ognjenns) [reply RestCommClient/Classes/Push/Binding.h:26]: yes, but we dont use anything from it, except the sid's once they are send to server.

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (ognjenns) [reply Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm:223]: ok

File: RestCommClient/Classes/Push/Binding.h
Comment (atsakiridis) [reply RestCommClient/Classes/Push/Binding.h:26]: Ok let me check the whole thing in the second-pass review

File: RestCommClient/Classes/Push/PushApiManager.m
Comment (atsakiridis) [inline RestCommClient/Classes/Push/PushApiManager.m]: Please check typo

File: RestCommClient/Classes/RCRegisterPushDelegate.h
Comment (atsakiridis) [inline RestCommClient/Classes/RCRegisterPushDelegate.h]: Can we keep just one callback for both success & failure scenarios? It allows for less boilerplate code and also is the direction we have been taking on Android as well

**Review by atsakiridis — CHANGES_REQUESTED** [review]

File: RestCommClient/Classes/RCRegisterPushDelegate.h
Comment (ognjenns) [reply RestCommClient/Classes/RCRegisterPushDelegate.h]: Sure. Will have one method.

File: RestCommClient/Classes/Push/PushApiManager.m
Comment (ognjenns) [reply RestCommClient/Classes/Push/PushApiManager.m]: will fix.

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (atsakiridis) [inline Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm]: Let's avoid the prefix rc, as we haven't used it in any other RCConnection/RCDevice callback. Remember this is public API and we strive for consistency ;)

Maybe just registeredForPush or pushDidRegister?

**Review by atsakiridis — CHANGES_REQUESTED** [review]

File: Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm
Comment (ognjenns) [reply Examples/restcomm-olympus-nopod/restcomm-olympus/MainTableViewController.mm]: registeredForPush it is.

**Review by atsakiridis — APPROVED** [review]

**Pull request merged**
